File: server/app/simulation.py
"""Movement simulation.

This is the function the client re-implements for prediction
(client/src/game/simulation.ts). Any change here MUST be mirrored there or
prediction will drift and the player will rubber-band.

Movement is continuous: velocity * dt in world pixels, then one axis-separated
collision pass against the tile grid. Cost per entity is constant — a couple of
float ops plus an overlap test over at most 2x2 tiles.
"""

from __future__ import annotations

from .config import (
    CARRY_MAX_WEIGHT,
    CARRY_SLOW_AT_MAX,
    CARRY_SLOW_FLOOR,
    CARRY_SLOW_START,
    HIT_STAGGER_SCALE,
    MOVE_SPEED,
    PLAYER_HALF_HEIGHT,
    PLAYER_HALF_WIDTH,
    SPRINT_SPEED,
    STAMINA_DRAIN,
    STAMINA_MAX,
    STAMINA_RECOVER,
    STAMINA_REGEN_REST,
    STAMINA_REGEN_WALK,
)
from .entities import InputCmd, Player
from .world import TileMap

_SQRT1_2 = 0.7071067811865476


def carry_scale(weight: float, max_weight: float = CARRY_MAX_WEIGHT) -> float:
    """How much of MOVE_SPEED a body gets at this carried weight.

    Full speed up to CARRY_SLOW_START of max. Then a straight line down to
    CARRY_SLOW_AT_MAX at the cap, and it keeps falling if they go over,
    never below CARRY_SLOW_FLOOR. Mirror: client/src/game/simulation.ts.

    `max_weight` is a PARAMETER rather than the constant because a skill moves
    it (`skills.Mods.carry`). The shape of the curve is still one decision —
    only where the free band ends slides.
    """
    if max_weight <= 0.0:
        return 1.0
    ratio = weight / max_weight
    if ratio <= CARRY_SLOW_START:
        return 1.0
    span = 1.0 - CARRY_SLOW_START
    t = (ratio - CARRY_SLOW_START) / span if span > 0.0 else 1.0
    scale = 1.0 + (CARRY_SLOW_AT_MAX - 1.0) * t
    return CARRY_SLOW_FLOOR if scale < CARRY_SLOW_FLOOR else scale


def move_dir(cmd: InputCmd) -> tuple[float, float]:
    dx = (1.0 if cmd.right else 0.0) - (1.0 if cmd.left else 0.0)
    dy = (1.0 if cmd.down else 0.0) - (1.0 if cmd.up else 0.0)
    if dx != 0.0 and dy != 0.0:
        dx *= _SQRT1_2
        dy *= _SQRT1_2
    return dx, dy


def running(player: Player, cmd: InputCmd, moving: bool) -> bool:
    """Whether this body is actually RUNNING this tick.

    Holding SHIFT is a request, not a state: a body standing still is not
    running (it would drain the bar for nothing), and a body that spent the bar
    is locked out until `STAMINA_RECOVER` of it is back — see `Player.winded`.
    Mirror: client/src/game/simulation.ts.
    """
    return moving and cmd.sprint and not player.winded and player.stamina > 0.0


def step_stamina(player: Player, run: bool, moving: bool, dt: float) -> None:
    """Spend or refill the breath, and work the exhaustion latch.

    STATELESS BY DESIGN — no rest timer, no cooldown. What the bar reads is
    decided entirely by (running, moving), so prediction can replay it from the
    server's number and land on exactly the value the server holds. The one
    piece of memory is `winded`, and it is a latch rather than a clock, which
    is why it rides the snapshot beside the number.

    Standing still refills faster than walking does. That is the only place the
    system asks for a decision: catching your breath properly means stopping in
    a dark forest. Mirror: client/src/game/simulation.ts.
    """
    if run:
        player.stamina -= STAMINA_DRAIN * dt
        if player.stamina <= 0.0:
            player.stamina = 0.0
            player.winded = True
        return
    regen = STAMINA_REGEN_WALK if moving else STAMINA_REGEN_REST
    player.stamina = min(STAMINA_MAX, player.stamina + regen * dt)
    if player.winded and player.stamina >= STAMINA_MAX * STAMINA_RECOVER:
        player.winded = False


def step_stagger(player: Player, dt: float) -> float:
    """Run the drag from the last blow down, and return what it multiplies by.

    A CLOCK, TICKED IN THE WALK, and it lives here rather than in the room's
    tick for exactly one reason: prediction. The client replays unacked inputs
    through `apply_input` after every reconcile, so anything that decays with
    time and changes speed has to decay inside the same function, or the replay
    walks the body at a speed the server never used. Same shape as
    `step_stamina` above and for the same reason.

    Mirror: client/src/game/simulation.ts.
    """
    if player.stagger <= 0.0:
        return 1.0
    player.stagger = max(0.0, player.stagger - dt)
    return HIT_STAGGER_SCALE


def apply_input(player: Player, cmd: InputCmd, world: TileMap, dt: float) -> None:
    dx, dy = move_dir(cmd)
    moving = dx != 0.0 or dy != 0.0
    run = running(player, cmd, moving)
    step_stamina(player, run, moving, dt)

    mods = player.skills.mods
    speed = MOVE_SPEED * mods.speed * carry_scale(player.carry_weight, mods.carry)
    if run:
        speed *= SPRINT_SPEED
    # THE SHIELD IS THE LAST TERM AND IT MULTIPLIES EVERYTHING. A body behind
    # one is slow whatever else is true about it — sprinting behind a riot
    # shield is still slower than walking without one, which is the whole
    # reason raising it is a decision rather than a posture. Resolved before
    # this runs; see `Player.block_speed`.
    speed *= player.block_speed
    # AND THE DRAG IS THE LAST TERM OF ALL, under even the shield. Being hit
    # takes precedence over every choice the player made about how fast to
    # move, because the entire point of it is that it is not a choice — a body
    # that could sprint out of a pack at full speed is a body for which being
    # surrounded costs nothing. See `HIT_STAGGER_SCALE`.
    speed *= step_stagger(player, dt)
    player.vx = dx * speed
    player.vy = dy * speed

    hw = PLAYER_HALF_WIDTH
    hh = PLAYER_HALF_HEIGHT
    player.x = world.move_axis(player.x, player.y, hw, hh, player.vx * dt, 0)
    player.y = world.move_axis(player.x, player.y, hw, hh, player.vy * dt, 1)

    player.aim_x = cmd.aim_x
    player.aim_y = cmd.aim_y

File: client/src/game/simulation.ts
/**
 * Movement simulation — mirror of server/app/simulation.py.
 *
 * This is the ONLY place the client is allowed to move the local player.
 * Prediction and reconciliation both call `applyInput`, so if this drifts from
 * the Python version the local player will visibly snap back.
 */

import type { GameConfig, InputPacket } from '../net/protocol';
import type { TileMap } from './world';

const SQRT1_2 = 0.7071067811865476;

export interface MovableState {
  x: number;
  y: number;
  vx: number;
  vy: number;
  ax: number;
  ay: number;
  /** Breath left, in the server's points. Mirror of `Player.stamina`. */
  stamina: number;
  /** Bar spent: SHIFT is refused until `staminaRecover` of it is back. */
  winded: boolean;
  /**
   * What the walk is multiplied by right now because of the shield. 1
   * whenever it is down. Mirror of `Player.block_speed`.
   *
   * A RESOLVED NUMBER RATHER THAN A LOOKUP, because this file is a
   * line-for-line mirror of `simulation.py` and movement code on either side
   * must not have to reach into a weapon catalog to know how fast a body is.
   * Both sides decide it in the same place — the frame the button is read —
   * and this just multiplies. See `Room.sync_block` and `Game.syncBlock`.
   */
  blockSpeed: number;
}

export function moveDir(input: InputPacket): { dx: number; dy: number } {
  let dx = (input.movement.right ? 1 : 0) - (input.movement.left ? 1 : 0);
  let dy = (input.movement.down ? 1 : 0) - (input.movement.up ? 1 : 0);
  if (dx !== 0 && dy !== 0) {
    dx *= SQRT1_2;
    dy *= SQRT1_2;
  }
  return { dx, dy };
}

/**
 * How much of `moveSpeed` a body gets at this carried weight.
 * Mirror of `carry_scale` in server/app/simulation.py.
 *
 * `maxWeight` overrides the config ceiling, because a skill moves it
 * (`skills.Mods.carry`, shipped on the roster as `mods.carry`). The curve is
 * still one decision; only where the free band ends slides.
 */
export function carryScale(
  weight: number,
  config: GameConfig,
  maxWeight?: number,
): number {
  const max = maxWeight ?? config.carryMaxWeight;
  const start = config.carrySlowStart;
  const atMax = config.carrySlowAtMax;
  const floor = config.carrySlowFloor;
  if (max <= 0) return 1;
  const ratio = weight / max;
  if (ratio <= start) return 1;
  const span = 1 - start;
  const t = span > 0 ? (ratio - start) / span : 1;
  const scale = 1 + (atMax - 1) * t;
  return scale < floor ? floor : scale;
}

/**
 * How heavy the walk *feels* past the free band. 0 at `carrySlowStart`,
 * 1 at max weight, and it keeps climbing if they go over. Footprints and
 * dust read this; speed uses `carryScale`.
 */
export function carryBurden(
  weight: number,
  config: GameConfig,
  maxWeight?: number,
): number {
  const max = maxWeight ?? config.carryMaxWeight;
  const start = config.carrySlowStart;
  if (max <= 0) return 0;
  const ratio = weight / max;
  const span = 1 - start;
  if (span <= 0) return ratio > start ? 1 : 0;
  return Math.max(0, (ratio - start) / span);
}

/**
 * Whether this body is actually RUNNING this tick.
 * Mirror of `running` in server/app/simulation.py.
 *
 * SHIFT is a request, not a state: a body standing still is not running, and a
 * body that spent the bar is locked out until it has recovered.
 */
export function isRunning(state: MovableState, input: InputPacket, moving: boolean): boolean {
  return moving && input.sprint && !state.winded && state.stamina > 0;
}

/**
 * Spend or refill the breath, and work the exhaustion latch.
 * Mirror of `step_stamina` in server/app/simulation.py.
 *
 * Stateless apart from the latch, which is what lets reconciliation replay it:
 * snap `stamina` / `winded` from the authoritative row, replay the pending
 * inputs through here, and the client lands on the number the server holds.
 */
export function stepStamina(
  state: MovableState,
  running: boolean,
  moving: boolean,
  config: GameConfig,
  dt: number,
): void {
  const max = config.staminaMax;
  if (running) {
    state.stamina -= config.staminaDrain * dt;
    if (state.stamina <= 0) {
      state.stamina = 0;
      state.winded = true;
    }
    return;
  }
  const regen = moving ? config.staminaRegenWalk : config.staminaRegenRest;
  state.stamina = Math.min(max, state.stamina + regen * dt);
  if (state.winded && state.stamina >= max * config.staminaRecover) {
    state.winded = false;
  }
}

export function applyInput(
  state: MovableState,
  input: InputPacket,
  world: TileMap,
  config: GameConfig,
  dt: number,
  weight = 0,
  /**
   * The owner's flattened skill mods. Mirror of `player.skills.mods` on the
   * server — a default of 1 / undefined is a body with no skills, which is
   * every body on the opening night.
   */
  mods?: { speed: number; carry: number },
): void {
  const { dx, dy } = moveDir(input);
  const moving = dx !== 0 || dy !== 0;
  const running = isRunning(state, input, moving);
  stepStamina(state, running, moving, config, dt);

  let speed = config.moveSpeed * (mods?.speed ?? 1) * carryScale(weight, config, mods?.carry);
  if (running) speed *= config.sprintSpeed;
  // THE SHIELD IS THE LAST TERM AND IT MULTIPLIES EVERYTHING. A body behind
  // one is slow whatever else is true about it — sprinting behind a riot
  // shield is still slower than walking without one, which is the whole
  // reason raising it is a decision rather than a posture.
  speed *= state.blockSpeed;
  state.vx = dx * speed;
  state.vy = dy * speed;

  const hw = config.playerHalfWidth;
  const hh = config.playerHalfHeight;
  state.x = world.moveAxis(state.x, state.y, hw, hh, state.vx * dt, 0);
  state.y = world.moveAxis(state.x, state.y, hw, hh, state.vy * dt, 1);

  state.ax = input.aim.x;
  state.ay = input.aim.y;
}
